jgrapht-core/src/main/java/org/jgrapht/alg/matching/blossom/v5/BlossomVState.java

```diff
@@ -0,0 +1,321 @@
+ */
+class BlossomVState<V, E> {
+    /**
+     * Number of nodes in the graph
+     */
+    final int nodeNum;
+    /**
+     * Number of edges in the graph
+     */
+    final int edgeNum;
+    /**
+     * The graph to search matching in
+     */
+    Graph<V, E> graph;
+    /**
+     * An array of nodes of the graph.
+     * BlossomVNode: the size of the array is nodeNum + 1. The node nodes[nodeNum] is an auxiliary node that is used
+     * as the first element in the linked list of tree roots
+     */
+    BlossomVNode[] nodes;
+    /**
+     * An array of edges of the graph
+     */
+    BlossomVEdge[] edges;
+    /**
+     * Helper variable to determine whether an augmentation has been performed
+     */
+    int treeNum;
+    /**
+     * Number of expanded blossoms
+     */
+    int removedNum;
+    /**
+     * Number of blossoms
+     */
+    int blossomNum;
+    /**
+     * Statistics of the algorithm performance
+     */
+    KolmogorovMinimumWeightPerfectMatching.Statistics statistics;
+    /**
+     * BlossomVOptions used to determine the strategies used in the algorithm
+     */
+    BlossomVOptions options;
+    /**
+     * Mapping from initial vertices to nodes
+     */
+    List<V> graphVertices;
+
+    List<E> graphEdges;
+
+    /**
+     * Constructs an initial algorithm's state
+     *
+     * @param graph         the graph to search matching in
+     * @param nodes         nodes used in the algorithm
+     * @param edges         edges used in the algorithm
+     * @param nodeNum       number of nodes in the graph
+     * @param edgeNum       number of edges in the graph
+     * @param treeNum       number of trees in the graph
+     * @param graphVertices generic vertices of the {@code graph} in the same order as nodes in {@code nodes}
+     * @param graphEdges    generic edges of the {@code graph} in the same order as edges in {@code edges}
+     * @param options       default or user defined options
+     */
+    public BlossomVState(Graph<V, E> graph, BlossomVNode[] nodes, BlossomVEdge[] edges,
+                         int nodeNum, int edgeNum, int treeNum,
+                         List<V> graphVertices, List<E> graphEdges, BlossomVOptions options) {
+        this.graph = graph;
+        this.nodes = nodes;
+        this.edges = edges;
+        this.nodeNum = nodeNum;
+        this.edgeNum = edgeNum;
+        this.treeNum = treeNum;
+        this.graphVertices = graphVertices;
+        this.graphEdges = graphEdges;
+        this.options = options;
+        this.statistics = new KolmogorovMinimumWeightPerfectMatching.Statistics();
+    }
+
+    /**
+     * Adds a new tree edge from {@code from} to {@code to}. Sets the to.currentEdge and to.currentDirection
+     * with respect to the tree {@code from}
+     *
+     * @param from the tail of the directed tree edge
+     * @param to   the head of the directed tree edge
+     */
+    public static BlossomVTreeEdge addTreeEdge(BlossomVTree from, BlossomVTree to) {
```

Comment:
Wouldn't this be more appropriate as a non-static method on BlossomVTree?  (Or static if you prefer for symmetry.)  Likewise for many of the other methods below; move them to the most appropriate class instead of here.